core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/XMLReaderBatchSource.java

```diff
@@ -0,0 +1,292 @@
+import org.apache.hadoop.mapreduce.Job;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * XML Reader Batch Source Plugin
+ * It is used to read XML files from HDFS with specified file properties and filters.
+ * It parses the read file into specified Output Schema.
+ * A {@link FileBatchSource} to use any distributed file system as a Source.
+ */
+@Plugin(type = BatchSource.PLUGIN_TYPE)
+@Name("XMLReader")
+@Description("Batch source for XML read from HDFS")
+public class XMLReaderBatchSource extends ReferenceBatchSource<LongWritable, Object, StructuredRecord> {
+  public static final Schema DEFAULT_XML_SCHEMA = Schema.recordOf(
+    "xmlSchema",
+    Schema.Field.of("offset", Schema.of(Schema.Type.LONG)),
+    Schema.Field.of("filename", Schema.of(Schema.Type.STRING)),
+    Schema.Field.of("record", Schema.of(Schema.Type.STRING))
+  );
+
+  private static final Gson GSON = new Gson();
+  private static final Type ARRAYLIST_PREPROCESSED_FILES  = new TypeToken<ArrayList<String>>() { }.getType();
+
+  private KeyValueTable processedFileTrackingTable;
+  private final XMLReaderConfig config;
+
+  public XMLReaderBatchSource(XMLReaderConfig config) {
+    super(config);
+    this.config = config;
+  }
+
+  @VisibleForTesting
+  XMLReaderConfig getConfig() {
+    return config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    config.validateConfig();
+    pipelineConfigurer.getStageConfigurer().setOutputSchema(DEFAULT_XML_SCHEMA);
+    pipelineConfigurer.createDataset(config.tableName, KeyValueTable.class.getName());
+  }
+
+  @Override
+  public void prepareRun(BatchSourceContext context) throws Exception {
+    Job job = JobUtils.createInstance();
+    Configuration conf = job.getConfiguration();
+    if (StringUtils.isNotEmpty(config.pattern)) {
+      conf.set(XMLInputFormat.XML_INPUTFORMAT_PATTERN, config.pattern);
+    }
+    XMLInputFormat.addInputPaths(job, config.path);
+    conf.set(XMLInputFormat.XML_INPUTFORMAT_REPROCESSING_REQUIRED, this.config.reprocessingRequired);
+    //set file tracking information in a temporary file, to be available to read outside plugin.
+    setFileTrackingInfo(context, conf);
+    if (StringUtils.isNotEmpty(this.config.actionAfterProcess)) {
+      conf.set(XMLInputFormat.XML_INPUTFORMAT_FILE_ACTION, this.config.actionAfterProcess);
+    }
+    if (StringUtils.isNotEmpty(this.config.targetFolder)) {
+      conf.set(XMLInputFormat.XML_INPUTFORMAT_TARGET_FOLDER, this.config.targetFolder);
+    }
+    conf.set(XMLInputFormat.XML_INPUTFORMAT_NODE_PATH, this.config.nodePath);
```

Comment:
nit - you use `this.config.nodePath` here, but `config.path` in the next line.
For simplicity, just omit the `this.`, since there's no local variable with the same name.
(same for other usages in this method)